File: 2019/2019W13/preppindataw13.R
library(readxl)
library(dplyr)
library(lubridate)

trans <- read_excel("E:/PD - Week 13.xlsx", 
                    sheet = "Transactions", col_types = c("numeric", 
                                                          "date", "numeric", "numeric"))
customer <- read_excel("E:/PD - Week 13.xlsx", 
                       sheet = "Customer Look-up")

temp <- trans %>%
  merge(., customer, by='Account') %>%
  mutate('Days Below Zero balance' = if_else(`Balance`<0,1,0),
         'Days Beyond Max Credit' = if_else(`Balance`*-1>`Max Credit`,1,0),
         'Week date' = if_else(floor_date(`Date`, 'weeks')<floor_date(`Date`, 'year'),
                               floor_date(`Date`, 'year'),floor_date(`Date`, 'weeks')),
         'Month date' = floor_date(`Date`, 'month'),
         'Quarter date' = floor_date(`Date`, 'quarter'),
         'Week' = as.integer(strftime(`Date`,format = '%U'))+1,
         'Month' = month(`Date`),
         'Quarter' = quarter(`Date`))

weekly <- temp %>%
  group_by(`Account`,`Name`,`Week`,`Week date`) %>% 
  summarise('Weekly Avg Transaction'= mean(`Transaction`),
            'Weekly Avg Balance'= mean(`Balance`),
            'Days Below Zero balance' = sum(`Days Below Zero balance`),
            'Days Beyond Max Credit' = sum(`Days Beyond Max Credit`)) %>%
  rename('Date'='Week date')
  
  
monthly <- temp %>%
  group_by(`Account`,`Name`,`Month`,`Month date`) %>% 
  summarise('Monthly Avg Transaction'= mean(`Transaction`),
            'Monthly Avg Balance'= mean(`Balance`),
            'Days Below Zero balance' = sum(`Days Below Zero balance`),
            'Days Beyond Max Credit' = sum(`Days Beyond Max Credit`)) %>%
  rename('Date'='Month date')

quarterly <- temp %>%
  group_by(`Account`,`Name`,`Quarter`,`Quarter date`) %>% 
  summarise('Quarterly Avg Transaction'= mean(`Transaction`),
            'Quarterly Avg Balance'= mean(`Balance`),
            'Days Below Zero balance' = sum(`Days Below Zero balance`),
            'Days Beyond Max Credit' = sum(`Days Beyond Max Credit`)) %>%
  rename('Date'='Quarter date')

File: 2019/2019W13/PreppindataW13.py
import pandas as pd
from datetime import timedelta

xls = pd.ExcelFile("E:\PD - Week 13.xlsx")
trans = pd.read_excel(xls,"Transactions") 
customer = pd.read_excel(xls,"Customer Look-up") 

temp = trans.merge(customer,on='Account', how='inner')
temp['Week']=temp.apply(lambda x: int(x['Date'].strftime("%U"))+1,axis=1)
temp['Week date']=temp.apply(lambda x: max(x['Date'].replace(month=1,day=1),  
    x['Date'] - timedelta(days=(x['Date'].weekday()+1)%7)).strftime("%Y/%m/%d"),axis=1)
temp['Month']=temp.apply(lambda x: x['Date'].month,axis=1)
temp['Month date']=temp.apply(lambda x: x['Date'].replace(day=1).strftime("%Y/%m/%d"),axis=1)
temp['Quarter']=temp.apply(lambda x: (x['Date'].month-1)//3+1,axis=1)
temp['Quarter date']=temp.apply(lambda x: x['Date'].replace(month=x['Quarter']*3-2,day=1).strftime("%Y/%m/%d"),axis=1)
temp['Days Below Zero balance'] = temp.apply(lambda x: 1 if x['Balance']<0 else 0,axis=1)
temp['Days Beyond Max Credit'] = temp.apply(lambda x: 1 if x['Balance']*-1>x['Max Credit'] else 0,axis=1)

agg_col = {'Transaction':'mean',
           'Balance':'mean',
           'Days Below Zero balance':'sum',
           'Days Beyond Max Credit':'sum'}

final = {}
for period in ['Week','Month','Quarter']:
    final[period] = temp.groupby(['Account','Name',period,period+' date'], as_index=False).agg(agg_col)
    final[period].rename(columns=
          {period+' date':'Date','Transaction':period+'ly Avg Transaction','Balance':period+'ly Avg Balance'},
          inplace=True)
